vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/InternalFieldBase.java

```diff
@@ -0,0 +1,423 @@
+import com.vaadin.flow.component.InputNotifier;
+import com.vaadin.flow.component.KeyNotifier;
+import com.vaadin.flow.component.shared.HasClearButton;
+import com.vaadin.flow.component.shared.HasClientValidation;
+import com.vaadin.flow.component.shared.HasTooltip;
+import com.vaadin.flow.data.binder.HasValidator;
+import com.vaadin.flow.data.value.HasValueChangeMode;
+import com.vaadin.flow.data.value.ValueChangeMode;
+import com.vaadin.flow.function.SerializableBiFunction;
+import com.vaadin.flow.function.SerializableFunction;
+
+/**
+ * Internal class that provides base functionality for input field components,
+ * such as {@link TextField}. Not intended to be used publicly.
+ *
+ * @param <TComponent>
+ *            Type of the component that extends from this class
+ * @param <TValue>
+ *            Type of the value of the extending component
+ */
+public abstract class InternalFieldBase<TComponent extends InternalFieldBase<TComponent, TValue>, TValue>
+        extends AbstractSinglePropertyField<TComponent, TValue>
+        implements CompositionNotifier, Focusable<TComponent>,
+        HasAutocapitalize, HasAutocomplete, HasAutocorrect, HasClearButton,
+        HasClientValidation, HasHelper, HasLabel, HasPrefixAndSuffix, HasSize,
+        HasStyle, HasTooltip, HasValidation, HasValidator<TValue>,
+        HasValueChangeMode, InputNotifier, KeyNotifier {
+
+    private ValueChangeMode currentMode;
+
+    private int valueChangeTimeout = DEFAULT_CHANGE_TIMEOUT;
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * The default value is {@link ValueChangeMode#ON_CHANGE}.
+     */
+    @Override
+    public ValueChangeMode getValueChangeMode() {
+        return currentMode;
+    }
+
+    @Override
+    public void setValueChangeMode(ValueChangeMode valueChangeMode) {
+        currentMode = valueChangeMode;
+        setSynchronizedEvent(
+                ValueChangeMode.eventForMode(valueChangeMode, "value-changed"));
+        applyChangeTimeout();
+    }
+
+    @Override
+    public void setValueChangeTimeout(int valueChangeTimeout) {
+        this.valueChangeTimeout = valueChangeTimeout;
+        applyChangeTimeout();
+    }
+
+    @Override
+    public int getValueChangeTimeout() {
+        return valueChangeTimeout;
+    }
+
+    private void applyChangeTimeout() {
+        ValueChangeMode.applyChangeTimeout(getValueChangeMode(),
+                getValueChangeTimeout(), getSynchronizationRegistration());
+    }
+
+    /**
+     * Sets the label for the field.
+     *
+     * @param label
+     *            value for the {@code label} property in the field
+     */
+    public void setLabel(String label) {
+        getElement().setProperty("label", label == null ? "" : label);
+    }
+
+    /**
+     * String used for the label element.
+     *
+     * @return the {@code label} property from the web component
+     */
+    public String getLabel() {
+        return getElement().getProperty("label");
+    }
```

Comment:
Since the class implements `HasLabel`, these shouldn't be needed?

Reply:
Checked the `HasLabel` interface implementation and there are two differences:

- [`setLabel(String)`](https://github.com/vaadin/flow/blob/a2fb35292cd0e5328cfd0ba0da32bd20ff378a99/flow-server/src/main/java/com/vaadin/flow/component/HasLabel.java#L38-L40) method doesn't take care of `null`,
- [`getLabel()`](https://github.com/vaadin/flow/blob/a2fb35292cd0e5328cfd0ba0da32bd20ff378a99/flow-server/src/main/java/com/vaadin/flow/component/HasLabel.java#L48-L51) returns `null` as a fallback for the label.

In case of `ComboBoxBase` we don't override these methods though 🤷‍♂️ 
@sissbruecker would you be OK to remove overrides from here too?

Reply:
I honestly can't remember. 

In this case, the `getLabel()` implementation is pretty much the same. And the `setLabel()` implementation here is not really consistent:
```
NumberField field = new NumberField();
System.out.println(field.getLabel()); => null

field.setLabel(null);
System.out.println(field.getLabel()); => ""
```

Besides, it would leave an empty `label` attribute on the client-side, not sure if that is necessary. If not, I'd say we can remove the overrides here?

Reply:
Thanks, good point about inconsistency, I didn't notice it myself. Will remove these overrides then.

Reply:
Removed.